crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1,32 +1,89 @@
 use rustc_hash::FxHashMap;
-use std::borrow::Cow;
-use std::sync::Arc;
+use salsa;
 
 use red_knot_module_resolver::{resolve_module, ModuleName};
 use ruff_db::files::File;
 use ruff_index::IndexVec;
 use ruff_python_ast as ast;
 use ruff_python_ast::{ExprContext, TypeParams};
 
-use crate::semantic_index::ast_ids::ScopedExpressionId;
-use crate::semantic_index::definition::{Definition, DefinitionNodeRef};
-use crate::semantic_index::symbol::{
-    FileScopeId, NodeWithScopeRef, ScopeId, ScopedSymbolId, SymbolTable,
+use crate::semantic_index::ast_ids::{HasScopedUseId, ScopedExpressionId};
+use crate::semantic_index::definition::{Definition, DefinitionKind};
+use crate::semantic_index::symbol::{NodeWithScopeRef, ScopeId};
+use crate::semantic_index::usedef::Expression;
+use crate::semantic_index::SemanticIndex;
+use crate::types::{
+    definitions_ty, use_def_map, ClassType, FunctionType, Name, Type, UnionTypeBuilder,
 };
-use crate::semantic_index::{symbol_table, SemanticIndex};
-use crate::types::{infer_types, ClassType, FunctionType, Name, Type, UnionTypeBuilder};
 use crate::Db;
+use ruff_db::parsed::parsed_module;
 
-/// The inferred types for a single scope.
+use crate::semantic_index::semantic_index;
+use crate::semantic_index::symbol::NodeWithScopeKind;
+
+/// Infer all types for a [`Definition`] (including sub-expressions).
+#[salsa::tracked(return_ref)]
+pub(crate) fn infer_definition_types<'db>(
+    db: &'db dyn Db,
+    definition: Definition<'db>,
+) -> TypeInference<'db> {
+    let _span = tracing::trace_span!("infer_definition_types", ?definition).entered();
+
+    let index = semantic_index(db, definition.file(db));
+
+    let mut builder = TypeInferenceBuilder::new(db, InferenceRegion::Definition(definition), index);
+
+    builder.infer_region();
+    builder.finish()
+}
```

Comment:
Without having looked at the code. Should `infer_region` consume `self` and directly return the result or should it/is it possible to infer multiple regions at once?

Reply:
Yeah, I think you're right; `infer_region()` and `finish()` should be the same method. Will fix.